---
title: "Using Seurat's manifold aligment with SWNE"
author: "Yan Wu"
date: "5/15/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(bitmapType="cairo")
```

This is a walkthrough demonstrating how to generate SWNE plots alongside the [Seurat](http://satijalab.org/seurat/) [manifold alignment pipeline](https://satijalab.org/seurat/v3.0/pancreas_integration_label_transfer.html) from three pancreas [datasets](https://www.dropbox.com/s/1zxbn92y5du9pu0/pancreas_v3_files.tar.gz?dl=1) generated using different single cell RNA-seq technologies.

To save time we will be using the pre-computed Seurat object `pancreas_integrated_seurat.Robj`, which can be downloaded [here](ftp://genome-miner.ucsd.edu/swne_files/multiple_pancreas_seurat_v3.RObj).


First let's load the required libraries
```{r message=FALSE,warning=FALSE}
library(Seurat)
library(swne)
```


Let's load the Seurat object
```{r}
se.obj <- readRDS("~/swne/Data/multiple_pancreas_seurat_v3.RObj")
```


Using the wrapper function `RunSWNE` is the simplest way to generate an SWNE embedding for an integrated Seurat object.
Just make sure that `DefaultAssay(object) == "integrated"`.
```{r}
genes.embed <- c("REG1A", "PPY", "SST", "GHRL", "VWF", "SOX10")
swne.embedding <- RunSWNE(se.obj, k = 20, genes.embed = genes.embed, snn.exp = 0.25, snn.k = 20)
```


We can make the SWNE plot
```{r}
clusters <- se.obj$celltype
PlotSWNE(swne.embedding, alpha.plot = 0.4, sample.groups = clusters, do.label = T, 
         label.size = 3.5, pt.size = 1.25, show.legend = F, seed = 42)
```


We also can show that there are no batch effects
```{r}
batch <- se.obj$tech
PlotSWNE(swne.embedding, alpha.plot = 0.4, sample.groups = batch, do.label = F, 
         label.size = 3.5, pt.size = 0.15, show.legend = T, seed = 42)
```


UMAP plot for comparison
```{r}
umap.emb <- Embeddings(se.obj, "umap")
PlotDims(umap.emb, sample.groups = clusters, show.legend = F, seed = 42,
         show.axes = F)
```
