---
title: "SWNE Walkthrough using Pagoda2"
author: "Yan Wu"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This is a quick walkthrough demonstrating how to generate SWNE plots alongside the [Pagoda2](https://github.com/hms-dbmi/pagoda2) pipeline using a [3k PBMC](https://support.10xgenomics.com/single-cell-gene-expression/datasets/1.1.0/pbmc3k) dataset as an example.

To save time we will be using the pre-computed Pagoda2 object `pbmc3k_pagoda2.Robj`, which can be downloaded [here](ftp://genome-miner.ucsd.edu/swne_files/pbmc3k_pagoda2.Robj).


First let's load the required libraries
```{r message=FALSE,warning=FALSE}
library(pagoda2)
library(swne)
library(Matrix)
```


Next let's load the Pagoda2 object
```{r}
p2 <- readRDS("/media/Home_Raid1/yan/swne/Data/pbmc3k_pagoda2.Robj")
```


Most scRNA-seq pipelines only use a subset of highly overdispersed genes for analysis. We'll pull out those variable genes here, as well as the cluster labels.
```{r}
## Pull out variable genes
n.od.genes <- 1.5e3
var.info <- p2$misc$varinfo; var.info <- var.info[order(var.info$lp, decreasing = F),];
od.genes <- rownames(var.info[1:n.od.genes,])
length(od.genes)

## Pull out clusters
clusters <- p2$clusters$PCA$multilevel
levels(clusters)
```


The easiest way to generate an SWNE embedding is to use the wrapper function `RunSWNE`
```{r}
## Run SWNE
genes.embed <- c("MS4A1", "GNLY", "CD3E", "CD14",
                 "FCER1A", "FCGR3A", "LYZ", "PPBP", "CD8A")
swne.embedding <- RunSWNE(p2, k = 14, var.genes = od.genes, genes.embed = genes.embed)

## Plot SWNE
PlotSWNE(swne.embedding, alpha.plot = 0.4, sample.groups = clusters,
         do.label = T, label.size = 3.5, pt.size = 1.5, show.legend = F,
         seed = 42)
```

Now we'll go through the SWNE embedding process step by step

First, let's pull out the counts, and log-scale and adjust the variance of each gene
```{r}
norm.counts <- ExtractNormCounts(p2, obj.type = "pagoda2", rescale = T, rescale.method = "log")
dim(norm.counts)
```


We use the `FindNumFactors` function to identify the optimal number of factors to use. This function can be slow for large datasets, since it iterates over different values of `k`, so a simple "hack" is to just set `k` equal to the number of significant principal components.
```{r}
k.range <- seq(2,16,2) ## Range of factors to iterate over
k.err <- FindNumFactors(norm.counts[od.genes,], k.range = k.range, n.cores = 8, do.plot = T)
```


We then run the NMF decomposition. We can initialize the NMF using either Independent Component Analysis (ica), Nonnegative SVD (nnsvd), or a completely random initialization. ICA is the best option for most datasets. The output of `RunNMF` is a list of the gene loadings (W) and NMF embedding (H).
```{r}
k <- 14
nmf.res <- RunNMF(norm.counts[od.genes,], k = k, init = "ica", n.cores = 8)
nmf.scores <- nmf.res$H
```


Compute the SNN matrix from the PCA embedding
```{r results=FALSE,message=FALSE,warning=FALSE}
p2$calculatePcaReduction(nPcs = 20, odgenes = od.genes)
pc.scores <- t(p2$reductions$PCA)
snn <- CalcSNN(pc.scores, k = 20, prune.SNN = 1/20)
```


Runs the SWNE embedding. The three key parameters are `alpha.exp`, `snn.exp`, and `n_pull`, which control how the factors and neighboring cells affect the cell coordinates.
```{r message=FALSE,warning=FALSE}
alpha.exp <- 1.25 # Increase this > 1.0 to move the cells closer to the factors. Values > 2 start to distort the data.
snn.exp <- 1.0 # Lower this < 1.0 to move similar cells closer to each other
n_pull <- 3 # The number of factors pulling on each cell. Must be at least 3.
swne.embedding <- EmbedSWNE(nmf.scores, snn, alpha.exp = alpha.exp, snn.exp = snn.exp,
                            n_pull = n_pull, proj.method = "umap", dist.use = "cosine")
```


For now, let's hide the factors by setting their names to the empty string `""`. We'll interpret them later
```{r}
swne.embedding$H.coords$name <- ""
```


To help with interpreting these cell clusters, let's pick some key PBMC genes to embed.
```{r}
genes.embed <- c("MS4A1", "GNLY", "CD3E", "CD14",
                 "FCER1A", "FCGR3A", "LYZ", "PPBP", "CD8A")
```


Since we only ran NMF on the overdispersed genes, we need to project the rest of the genes onto the NMF projection to get gene loadings for all genes.
```{r}
nmf.res$W <- ProjectFeatures(norm.counts, nmf.scores, n.cores = 8)
```


Now we can embed the key PBMC genes onto the visualization and remake the plot
```{r}
swne.embedding <- EmbedFeatures(swne.embedding, nmf.res$W, genes.embed, n_pull = 3)
```


Let's make the SWNE plot with the key genes embedded. The closer a cell or a cluster is to a gene, the higher the expression level. We set a seed for reproducible cluster colors, so that every plot will use the same colors to label the clusters.
```{r fig.width=5,fig.height=5}
PlotSWNE(swne.embedding, alpha.plot = 0.4, sample.groups = clusters, do.label = T,
         label.size = 3.5, pt.size = 1.5, show.legend = F, seed = 42)
```


We can validate this by overlaying the expression of one of these key genes onto the plot.
```{r fig.width=5,fig.height=5}
gene.use <- "CD8A"
gene.expr <- norm.counts[gene.use,]
FeaturePlotSWNE(swne.embedding, gene.expr, gene.use, alpha.plot = 0.4, label.size = 3.5, pt.size = 1.25)
```


We can also make a t-SNE plot for comparison.
```{r fig.width=5,fig.height=5}
tsne.scores <- p2$embeddings$PCA$tSNE
PlotDims(tsne.scores, sample.groups = clusters, pt.size = 0.75, label.size = 3.5, alpha = 0.3,
         show.legend = F, seed = 42, show.axes = F)
```


We can also interpret the factors by using the gene loadings matrix. Here, we extract the top 3 genes for each factor by gene loading. Since NMF tends to create a parts-based representation of the data, the factors often correspond to key biological processes or gene modules that explain the data.
```{r}
gene.loadings <- nmf.res$W
top.factor.genes.df <- SummarizeAssocFeatures(gene.loadings, features.return = 3)
head(top.factor.genes.df)
```


And finally, we can make a heatmap to visualize the top factors for each gene
```{r fig.width=4.0,fig.height=5.5}
gene.loadings.heat <- gene.loadings[unique(top.factor.genes.df$feature),]
ggHeat(gene.loadings.heat, clustering = "col")
```

Extract cluster colors for compatibility with other plotting methods (i.e. Monocle)
```{r}
color.mapping <- ExtractSWNEColors(swne.embedding, sample.groups = clusters, seed = 42)
color.mapping
```
